_posts/2022-04-26-fivetran-custom-connector.md

```diff
@@ -0,0 +1,335 @@
+- Automatic schema updates
+- Data de-deduplication
+- Destination ingestion optimisations
+- Logs and alerts to monitor events and troubleshoot issues
+
+Use FiveTran's Function connectors if:
+
+- Fivetran doesn't have a connector for your source
+- You are using private APIs or custom applications
+- You are using a source or API that Fivetran is unlikely to support in the near future
+- You want to sync unsupported file formats that require pre-processing
+- You have sensitive data that needs filtering or anonymizing before entering the destination
+
+*Fivetran* . Fivetran Custom Connector. (n.d.). Retrieved May 26, 2022, from https://fivetran.com/docs/functions 
+
+<br>
+
+***
+<br>
+## General Principles
+
+**State Management**
+
+State is a JSON object that contains cursors from the previous successful FiveTran executions run.
+
+[https://fivetran.com/docs/functions/faq/use-state-object](https://fivetran.com/docs/functions/faq/use-state-object) 
+
+[https://fivetran.com/docs/functions/faq/use-secrets-object](https://fivetran.com/docs/functions/faq/use-secrets-object)
+
+
+**Guardrails to consider**
+
+- Lambda timeouts and payload limit
+- Ensure calls to the Lambda are **idempotent** for a given FiveTran cursor
+- Be wary or avoid using cursors supplied by the pagination of an upstream API that may not be persistent.
+
+
+**FiveTran Pagination**
+
+FiveTran pagination allows your lambda function to specify if there is more data to be collected, this is achieved using the hasMore boolean return.
+
+When hasMore = True ,state is updated as normal, however FiveTran immediately calls the lambda with the updated state. This will keep occurring until hasMore = False, which then resets FiveTran to its default state.
+
+[https://fivetran.com/docs/functions/faq/use-hasmore-flag](https://fivetran.com/docs/functions/faq/use-hasmore-flag)
+
+Example Expected function payload
+
+```
+{
+      state: {
+        cursor: '2020-01-01',
+        paginationCounter: 0
+      },
+      secrets: Object
+    }
+```
+
+Example Expected Connector response format
+
+```
+state: {
+cursor: cursorPoint,
+paginationCounter: 1
+    },
+    insert: {
+      Table: apiResponseJson
+    },
+    schema: {
+      Table: {
+        primary_key: ['some_unique_key']
+      }
+    },
+    hasMore: boolean
+  }
+}
+
+```
+
+<br/>
+
+***
+
+<br/>
+## Sequence of FiveTran Connector Excecutions
+
+![](/img/five_tran_sequence.png)
+
+
+For the sake of this example and readability, the return JSON from the api call will be summarised as *apiJsonResponse*
+
+### Call 1: Initial Sync
+
+
+**FiveTranCall**
+
+```
+state: {}
+```
+
+**Connector Response**
+
+```
+{
+state: {
+lastUpdate: ''
+paginationCounter: 1
+},
+insert: {
+apiJsonResponseTable: apiJsonResponse
+},
+schema: {
+apiJsonResponseTable: {
+primary_key: ['id']
+}
+},
+hasMore: True
+}
+```
+
+Initial API sync with no state, api call is getting all data with no state and has returned a paginated response
```

Comment:
a bit clunky - [read it out loud](https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.2grqrue)

Reply:
Agreed

```suggestion
The initial FiveTran sync contains no state, triggering our custom connector to fetch all data from the data source API. The API has returned page 1/3 of a paginated response
```